### PR TITLE
Add pp formulanet

### PR DESCRIPTION
**Pull request by liuhongen1234567**
1. 添加了 PP-FormulaNet 公式识别算法
2. 在静态图下支持了UniMERNet 和 PP-FormulaNet的 KV cache 加速

### CODE REVIEW

File: tests/test.py
Comment (GreatV) [inline tests/test.py:1]: 这个test为什么只叫test

File: docs/algorithm/formula_recognition/algorithm_rec_ppformulanet.md [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+| PP-FormulaNet-L | Vary_VIT_B | [rec_pp_formulanet_l.yml](../../../configs/rec/PP-FormuaNet/rec_pp_formulanet_l.yml) |     0.9055   |     0.9206       | 0.9392  |     0.9273    | 0.9141 |     0.9213         |[训练模型](https://paddleocr.bj.bcebos.com/contribution/rec_ppformulanet_l_train.tar )|
+
+其中，SPE、CPE为UniMERNet的简单公式数据集和复杂公式数据集；Easy、Middle、Hard为PaddleX内部自建的简单公式数据集（LaTeX 代码长度 0-64）、中等公式数据集（LaTeX 代码长度  64-256）和复杂公式数据集（LaTeX 代码长度  256+）。
+
+## 2. 环境配置
+请先参考[《运行环境准备》](../../ppocr/environment.md)配置PaddleOCR运行环境，参考[《项目克隆》](../../ppocr/blog/clone.md)克隆项目代码。
+
+此外，需要安装额外的依赖：
+```shell
+apt-get install sudo
```
Comment (GreatV) [inline docs/algorithm/formula_recognition/algorithm_rec_ppformulanet.md:22]: 建议去掉这一句

**Review by GreatV — COMMENTED** [review]
建议把新增的模型文档link上
https://github.com/PaddlePaddle/PaddleOCR/blob/0697d248f8d56d004ea33216004c5238f82ef22e/mkdocs.yml#L374

**Comment by liuhongen1234567** [comment]
> 建议把新增的模型文档link上
> 
> https://github.com/PaddlePaddle/PaddleOCR/blob/0697d248f8d56d004ea33216004c5238f82ef22e/mkdocs.yml#L374

好的，已添加

**Comment by liuhongen1234567** [comment]
@GreatV 辛苦在review 一下呢？

File: configs/rec/PP-FormuaNet/rec_pp_formulanet_l.yml [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+  eval_batch_step: [0,  417 ]
+  cal_metric_during_train: True
+  pretrained_model:
+  checkpoints:
+  save_inference_dir:
+  use_visualdl: False
+  infer_img: doc/datasets/pme_demo/0000013.png
+  infer_mode: False
+  use_space_char: False
+  rec_char_dict_path:  ppocr/utils/dict/unimernet_tokenizer
```
Comment (GreatV) [inline configs/rec/PP-FormuaNet/rec_pp_formulanet_l.yml:18]: rec_char_dict_path在配置文件中重复出现了多次:

```yaml
# Global部分
rec_char_dict_path:  ppocr/utils/dict/unimernet_tokenizer

# PostProcess部分
PostProcess:
  name:  UniMERNetDecode
  rec_char_dict_path: ppocr/utils/dict/unimernet_tokenizer

# Train部分的transforms
  transforms:
      - UniMERNetLabelEncode:
          rec_char_dict_path: ppocr/utils/dict/unimernet_tokenizer

# Eval部分的transforms  
      - UniMERNetLabelEncode:
          rec_char_dict_path: ppocr/utils/dict/unimernet_tokenizer
```

建议:
1. 将rec_char_dict_path Global 部分作为全局配置
2. 其他模块通过引用Global中的配置值来使用,避免重复定义
3. 修改后的示例结构:

```yaml
Global:
  rec_char_dict_path: ppocr/utils/dict/unimernet_tokenizer
  # ... 其他全局配置

PostProcess:
  name: UniMERNetDecode
  rec_char_dict_path: ${Global.rec_char_dict_path}

Train:
  transforms:
    - UniMERNetLabelEncode:
        rec_char_dict_path: ${Global.rec_char_dict_path}

Eval:
  transforms:
    - UniMERNetLabelEncode:
        rec_char_dict_path: ${Global.rec_char_dict_path}
```

File: configs/rec/PP-FormuaNet/rec_pp_formulanet_s.yml
Comment (GreatV) [inline configs/rec/PP-FormuaNet/rec_pp_formulanet_s.yml:1]: 同上，建议其他模块通过引用Global中的配置值来使用,避免重复定义

File: ppocr/losses/rec_ppformulanet_loss.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    """
+    PPFormulaNet_L adopt CrossEntropyLoss for network training.
+    """
+
+    def __init__(self, vocab_size=50000):
+        super(PPFormulaNet_L_Loss, self).__init__()
+        self.ignore_index = -100
+        self.vocab_size = vocab_size
+        self.pad_token_id = 1
+        # 训练时是否忽略 padding？
```
Comment (GreatV) [inline ppocr/losses/rec_ppformulanet_loss.py:57]: 这里的注释最好改为英文

**Review by GreatV — APPROVED** [review]
LGTM

**Comment by Bestboy125** [comment]
请问这部分公式推理速度怎么样呢，paddle_ocr_v4可以快速完成一张图片多个公式检测再识别（原理是否是并行识别呢），是否有一种办法将这个模型，跟检测模型集成，然后并行识别呢，从而加快复杂图片的推理速度呢

**Comment by liuhongen1234567** [comment]
1. 关于速度方面，PP-FormulaNet-S 单图推理大概在 200 ms 左右，组batch=15可以达到 30 ms ；PP-FormulaNet-L 单图推理大概在 2000 ms 左右，组batch=15可以达到 300 ms ；
2.  paddle_ocr_v4可以快速完成一张图片多个公式检测再识别方面，公式识别模型（PP-FormulaNet-S 、PP-FormulaNet-L和 UniMERNet）是可以支持并行识别的，但是ppocrv4目前在公式检测方面效果是否ok不确定
3. 是否有一种办法将这个模型，跟检测模型集成，然后并行识别呢。这个是可以实现的，不过这种方案目前集成在PaddleX中，可以参考 https://github.com/PaddlePaddle/PaddleX/blob/release/3.0-beta2/docs/pipeline_usage/tutorials/ocr_pipelines/formula_recognition.md。 不过这种方案是几个月前做的，仅支持了 laTeX-OCR （不支持并行）；预计下周会在paddleX develop版本接入最新的公式识别产线，从而支持最新的几个模型

**Comment by WZJAI2018** [comment]
你好，我尝试将PP-FormulaNet-S的模型转换为onnx的模型，我用的是paddle自带的工具，我运行的命令是:
paddle2onnx --model_dir /root/.paddlex/official_models/PP-FormulaNet-S/ --model_filename inference.json --params_filename inference.pdiparams --save_file PP-FormulaNet-S.onnx --opset_version 11  --enable_onnx_checker True
但是转换出来的onnx模型，我在python环境测试了一下，同一张图，onnx模型的输出和paddle原生模型的输出有点不一致。

onnx模型的输出：
[array([[   0,    0,    0,   82, 2186,  243,  113,  279,  243,  115,  243,
          82,  243,  243,  243,   82,  243,  243,   82,  243,   82,  243,
         243,   30,   82,   82,   82,  707,  243,  113,  243,   39,  115,
         243,  113,  243,   82,  243,  115,  243,   82, 2387,  243,   82,
         243,   30,  243,  243,  707,   82,   82, 1278,  113,  113,  243,
         115,  115,  243,  113,  243,   40,  243,  115,  243,   31,    2]],
      dtype=int64)]
paddle原生模型的输出:
[array([[   
         0,    0,    0,   82, 2186,  243,  113,  243,   82,  707,  243,
         113,  243,   40,  243,  115,  243,  113,  243,   82, 1278,  243,
         115,  243,  115,  243,   82, 2387,  243,   30,  279,  243,   35,
         243,   82,  707,  243,  113,  243,   82, 1278,  243,  115,  243,
         113,  243,   42,  243,  115,  243,   31,    2,   82,    2]])]
我在paddlex的源码里面，在模型推理之前把模型的输入数据存到txt文档里，然后onnx也是在模型推理之前把模型输入存到txt文档里，对面了一下这两个txt文档的内容是一致的，我怀疑是paddle2onnx转出来的onnx模型存在问题，所以我想请教一下，应该怎么正确的导出onnx模型
